### PR TITLE
ENH: Make stirling2 support pure python when overflow

### PR DESCRIPTION
**Pull request by steppi**
This PR makes stirling2 follow a similar api to `scipy.special.comb`. It now has a keyword argument `exact`. When True, it uses exact integer arithmetic, but only accepts scalar integer arguments, not arrays. It first tries to use the ufunc based on `stirling2` in `cephes/stirling.c`. If overflow occurs, it tries again with a pure python implementation.  This ufunc has been renamed to `_stirling2` and is now private.

This PR also adds`stirling2_inexact` to `cephes/stirling.c` and a private ufunc based on it. This is used when `exact=False`. So far it just uses `stirling2` from `stirling.c` and casts to double if no overflow, otherwise when overflow it returns `nan`. This is just a placeholder. The plan is to use the uniformly valid asymptotic approximation found on [Wikipedia](https://en.wikipedia.org/wiki/Stirling_numbers_of_the_second_kind#Asymptotic_approximation) which was developed by [Nico Temme](https://onlinelibrary.wiley.com/doi/abs/10.1002/sapm1993893233). 

There is also a new test for the `exact=False` case and `test_modulus` has been given new test cases. I love the idea of `test_modulus`, but the small values previously tested could just as easily be tested exactly based on a reference table. I think this test is better as a sanity check that the function is working for large Stirling numbers.

### CODE REVIEW

File: scipy/special/_stirling2.py
Comment (rlucas7) [inline scipy/special/_stirling2.py]:
```suggestion
        return 1 if n == 0 else 0
```

File: scipy/special/cephes/stirling.c
Comment (rlucas7) [inline scipy/special/cephes/stirling.c]: I fine with the idea of an approximation but I'd prefer they exist as separate PRs. 
In other words PR 1 contains code in python to use backend stirling2 and PR 2 implements the Temme approximation. 

This is so that the enhancements are-to some extent independent-the PR1 can merge and work fine while PR 2 is in development. 

Also, wouldn't this be 'private' too? e.g. (`_stirling2_inexact`) ?

File: scipy/special/tests/test_stirling.py
Comment (rlucas7) [inline scipy/special/tests/test_stirling.py]:
```suggestion
    """Test triangle of values for 0 <= n <= 10 and 0 <= k <= n
```

not to be pedantic but "full" could mean `-inf < k < inf` to some people

File: scipy/special/_stirling2.py
Comment (rlucas7) [inline scipy/special/_stirling2.py:48]: Also, there are potential inputs where all choices of `n` with `1<k<n` will be overflow
so it seems like we'd want some sort of input validation method here to save the compute in `stirling2`.

File: scipy/special/_stirling2.py
Comment (rlucas7) [inline scipy/special/_stirling2.py]: This begs the question: why have the `c` implementation in the first place? (e.g. does it really save that much time that the additional complexity is worth the effort?) 

A waste here is that the `_stirling2()` function does a bunch of computation that is ignored 
and then *redone* in the `_stirling2_pyint()` call. 

Couldn't we reuse some of that preliminary calculation?

**Review by rlucas7 — CHANGES_REQUESTED** [review]
left some comments

File: scipy/special/cephes/stirling.c
Comment (steppi) [reply scipy/special/cephes/stirling.c]: Yeah, agreed. I wanted to add the `exact` keyword arg now even though `exact=False` is not that useful yet, just so we wouldn't have to change the api in the future. I'll follow up with a Temme PR once we get your PR merged into main. 

And yes, that should be private. I hacked this together hastily this morning.

File: scipy/special/tests/test_stirling.py
Comment (steppi) [reply scipy/special/tests/test_stirling.py]: Not pedantic. You're right

File: scipy/special/_stirling2.py
Comment (steppi) [reply scipy/special/_stirling2.py:48]: If `n` and `k` are restricted to 32 bit ints then ${n \brace n-1} = {n \choose 2}$ will never exceed the max 64 bit int, but that just means the condition for the check needs to be adjusted a little. This will be rendered moot by something you said in another comment.

File: scipy/special/_stirling2.py
Comment (steppi) [reply scipy/special/_stirling2.py]: Yeah. For `exact=True` I don't think we need the C function. We really just need one for `exact=False` where `stirling2` will be a ufunc. I think we can just expose inexact from cephes with signature double f(long, long) as a private ufunc. ~When it has the temme approximation, it might be useful as public~ nevermind, it's better to have it just accessible through stirling2. The situation with binom and comb was kind of confusing. 

We could reuse the work from C if we use a C extension function that returns a Python list of the last array that didn't overflow, along with its index. We'd want to use Cython to make working with the Python API easier. I don't think it's necessary though. As a scalar function, the pure Python version is fast enough.

Comment (steppi) [reply scipy/special/_stirling2.py]: Some data: I did a very simple benchmark.

```python

from scipy.special._stirling2 import stirling2, _stirling2_pyint


def check_c():
    for n in range(26):
        for k in range(n + 1):
            stirling2(n, k)


def check_py():
    for n in range(26):
        for k in range(n + 1):
            _stirling2_pyint(n, k)


def check_inexact():
    for n in range(26):
        stirling2(n, np.arange(n + 1), exact=False)
```

```
In [2]: %timeit check_c()
764 µs ± 799 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [3]: %timeit check_py()
2.52 ms ± 14.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit check_inexact()
134 µs ± 181 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Yeah, I don't think it's worth the added complexity to call C for `exact=True`, but the ufunc for exact=False does give a nice speedup.

**Comment by steppi** [comment]
I made `stirling2` just use the python implementation when `exact=True`. There was a 3x speedup with C for smaller values, but it's probably not worth it. It's funny, the Stirling numbers which don't overflow a long could fit in a small lookup table, except for the $\{n \brace n - 1}$ case which has a simple closed form anyway. I don't know why I was worried about micro-optimizing there. The interesting case is obviously for large values.

 I also made it so that only the inexact stirling2 function is exposed as a private ufunc. I changed it's name to `stirling2_approx` which I think sounds better.

**Comment by steppi** [comment]
I've made the function `stirling2` in `stirling.c` use `uint64_t` for the result since this function is now private and we don't need to worry about returning to Python from this function. This buys us a little more room before overflow. I also updated some of the documentation in response to the latest changes. In addition ${n \brace n - 1}$ now uses the simple closed form ${n \choose 2}$.

At this point, after a little more polishing for the docstring of `stirling2`  in `_stirling2.py` I think I'd be ready to accept your PR once this once this PR gets merged into your branch, please give this PR another look first though to make sure I didn't miss anything.

File: scipy/special/cephes/stirling.c
Comment (steppi) [reply scipy/special/cephes/stirling.c]: Oh, I see what you were saying. I thought you were talking about a Python function, but I had made that private already. In C, leading underscores are reserved for functions and types used by the compiler and standard library. The convention is different than in Python. See [here](https://stackoverflow.com/questions/39625352/why-do-some-functions-in-c-have-an-underscore-prefix)

**Comment by steppi** [comment]
Oh, there's a bug now for `exact=False`. Working on it.

```
In [3]: stirling2(100, 2)
Out[3]: 633825300114114700748351602687

In [4]: stirling2(100, 2, exact=False)
Out[4]: 1.8446744073709552e+19
```

**Comment by steppi** [comment]
Better. I forgot to change the return values after going to an unsigned type.

**Comment by steppi** [comment]
@rlucas7, would you prefer I update the docstring here and you review, or you merge this when you think it's ready, and then you update the docstring and I review it?

**Comment by rlucas7** [comment]
> 

I'm not following the two options ... can you elaborate?

File: scipy/special/cephes/stirling.c
Comment (rlucas7) [inline scipy/special/cephes/stirling.c:31]: I'm not sure we can support these in special at the moment, they would be `long long` correct?

**Review by rlucas7 — COMMENTED** [review]
My suggestion is to takethe `exact code from here and put that into the PR I opened that you were reviewing https://github.com/scipy/scipy/pull/18103 

We can add the `exact=False` branch after the fact using the default `exact=True` which would be the default from pr-18103 anyway. Unless we think that it's worthwhile to make `exact` a required argument. I have no strong opinions either way e.g. I'm fine having 18103 merged sooner or merged later (if we want to make `exact` a required argument. 

Thoughts?

File: scipy/special/cephes/stirling.c
Comment (steppi) [reply scipy/special/cephes/stirling.c:31]: The way this is set up, this function is completely internal to this file, note the `static`. It's not exposed in special. `stirling2_approx` uses this function internally but casts the result to double, which is handled. See [here](https://github.com/scipy/scipy/blob/0a017f2f79cfd27f7d7d59a59ed759b00174f737/scipy/special/_hyp2f1.pxd#L225-L233) for an example where I used `uint64_t` in an internal function which is used in the kernel of a ufunc.

**Comment by steppi** [comment]
> My suggestion is to takethe `exact code from here and put that into the PR I opened that you were reviewing [scipy#18103](https://github.com/scipy/scipy/pull/18103)
> 
> We can add the `exact=False` branch after the fact using the default `exact=True` which would be the default from pr-18103 anyway. Unless we think that it's worthwhile to make `exact` a required argument. I have no strong opinions either way e.g. I'm fine having 18103 merged sooner or merged later (if we want to make `exact` a required argument.
> 
> Thoughts?

My thought is that we should keep the API consistent from the start, rather than adding an `exact` keyword arg later. I've implemented the asymptotic expansions and can make a PR for that once we get yours merged, but it may be a while before it can be properly reviewed. In the meantime, I think it would be good if  there's still an `exact=False` case. Even without asymptotic expansions, the `exact=False` case is still useful since it's a ufunc rather than scalar valued, and gives a significant speedup when there's no overflow.

**Comment by steppi** [comment]
> > 
> 
> I'm not following the two options ... can you elaborate?

Sure, sorry for not explaining clearly.  I've submitted this PR to your PR branch as a way to make suggestions that can't be made with just the github UI. If you accept the updates here into your branch, the docstring you wrote will no longer be consistent with `stirling2`'s behavior. If you'd be happy to merge some or all of the changes from my branch into your PR branch then either

1. I can update the docstring in my branch before you merge this PR (or a subset of it) into your branch.
2. You can merge my branch into yours with the now inconsistent docstring and then work on the docstring yourself.

In the first case, you would review my updates to the docstring before accepting my PR for merging into your branch. In the latter case, I would review your updates to the docstring before accepting your PR.

Do you think the changes in my branch are worth integrating into your branch? I think a pure Python implementation makes the most sense because Stirling numbers overflow fixed width ints so quickly, but eventually, the asymptotic expansions will make `exact=False` pretty useful as a way to compute many Stirling numbers quickly. I think it would be best if we keep the API consistent from the start, so that's why I think it's best to add `exact=False` now.

**Comment by rlucas7** [comment]
There's two different things being discussed here so I'll separate them out:

### **Sequencing/separation of the exact and approximate/inexact implementations:**

> My thought is that we should keep the API consistent from the start, rather than adding an `exact` keyword arg later.

That's why I'm suggesting to add the `exact=True`, e.g. exact case executed by default. This ***does*** keep the signature the same when called e.g. `stirling(n=3,k=2)` performs the same as `stirling(n=3,k=2, exact=True)` if the `exact` argument is added with the Temme approximation pr and the ***default*** value is set to `True`.  

> I've implemented the asymptotic expansions and can make a PR for that once we get yours merged, but it may be a while before it can be properly reviewed. In the meantime, I think it would be good if there's still an `exact=False` case. Even without asymptotic expansions, the `exact=False` case is still useful since it's a ufunc rather than scalar valued, and gives a significant speedup when there's no overflow.

I'm fine with adding this once there's code to execute that codepath what I'm suggesting by adding the argument `exact=True` to `stirling2(n, k)` with the PR that adds the approximation code does this exactly (pun intended). :) 

### **Ufunc/c implementation and python wrapper versus pure python:**

>Do you think the changes in my branch are worth integrating into your branch? I think a pure Python implementation makes the most sense because Stirling numbers overflow fixed width ints so quickly, 

Yes I think those should go directly with the PR that implements the `exact` case.

However, @rgommers was suggesting the ufunc implementation in the issue [here](https://github.com/scipy/scipy/issues/17890#issuecomment-1410100391) so let's get his input. Ralf, after looking at this we think the pure python approach makes more sense because of how quickly overflow happens with Stirling numbers of the second kind and the speedup appears to be not that much w/C once we use the memory optimization @steppi invented. In light of this do you still think a ufunc in C is desired or are you ok w/a pure python implementation?

For added context since the discussion on the issue, the inclusion-exclusion (IE) approach overflows much quicker than a dynamic programming approach so we're not implementing an IE code path.

File: scipy/special/cephes/stirling.c
Comment (rlucas7) [reply scipy/special/cephes/stirling.c:31]: > The way this is set up, this function is completely internal to this file, note the `static`. It's not exposed in special. `stirling2_approx` uses this function internally but casts the result to double, which is handled. See [here](https://github.com/scipy/scipy/blob/0a017f2f79cfd27f7d7d59a59ed759b00174f737/scipy/special/_hyp2f1.pxd#L225-L233) for an example where I used `uint64_t` in an internal function which is used in the kernel of a ufunc.

ok. I understand now, thanks for clarifying.

File: scipy/special/cephes/stirling.c
Comment (rlucas7) [reply scipy/special/cephes/stirling.c]: > Yeah, agreed. I wanted to add the `exact` keyword arg now even though `exact=False` is not that useful yet, just so we wouldn't have to change the api in the future. I'll follow up with a Temme PR once we get your PR merged into main.
> 
> And yes, that should be private. I hacked this together hastily this morning.

See my detailed comment at bottom of thread. TL;DR I don't think we should add `exact` argument until we have 2 distinct values (`exact=True` and `exact=False`) that have working codepaths. We can do the `exact=True` as the default codepath ***without*** the `exact` argument at all.

File: scipy/special/cephes/stirling.c
Comment (rlucas7) [inline scipy/special/cephes/stirling.c:120]:
```suggestion
    if (k < 0 || k > n || n < 0){
```
because of the added check above (n==0 && k == 0)

Comment (rlucas7) [inline scipy/special/cephes/stirling.c:120]:
```suggestion
    if (k < 0 || k > n || n < 0){
```
because of the check added above (n==0 && k==0)

**Comment by steppi** [comment]
> That's why I'm suggesting to add the `exact=True`, e.g. exact case executed by default. This _**does**_ keep the signature the same when called e.g. `stirling(n=3,k=2)` performs the same as `stirling(n=3,k=2, exact=True)` if the `exact` argument is added with the Temme approximation pr and the _**default**_ value is set to `True`.
> 
> > I've implemented the asymptotic expansions and can make a PR for that once we get yours merged, but it may be a while before it can be properly reviewed. In the meantime, I think it would be good if there's still an `exact=False` case. Even without asymptotic expansions, the `exact=False` case is still useful since it's a ufunc rather than scalar valued, and gives a significant speedup when there's no overflow.
> 
> I'm fine with adding this once there's code to execute that codepath what I'm suggesting by adding the argument `exact=True` to `stirling2(n, k)` with the PR that adds the approximation code does this exactly (pun intended). :)

OK, I think that's fine. I've checked that `factorialk` has had an `exact` arg which raises `NotImplementedError` since 2002, and couldn't find the rationale behind this. It is backwards compatible without the arg, so it's fine. It really only serves the purpose of documenting that we know there should be an `exact` arg.



> However, @rgommers was suggesting the ufunc implementation in the issue [here](https://github.com/scipy/scipy/issues/17890#issuecomment-1410100391) so let's get his input. Ralf, after looking at this we think the pure python approach makes more sense because of how quickly overflow happens with Stirling numbers of the second kind and the speedup appears to be not that much w/C once we use the memory optimization @steppi invented. In light of this do you still think a ufunc in C is desired or are you ok w/a pure python implementation?

I think a ufunc with fixed width int is not too useful. The values overflow too quickly. The platonic state of `stirling2` would be having an implementation which works with arbitrary precision ints for `exact=True`, and to have a ufunc for `exact=False` which returns a double which approximates `stirling2` within 1 ulp of machine precision. I don't think we'll actually get to ` ulp` when I PR that later, probably around 100, but that will still be useful.

@rlucas, I'm going to close this PR. My suggestion is to update your PR to include only the pure Python implementation and strip out the C stuff. I'll try to have a PR for the asymptotic approximation ready within 2 weeks from when we get yours merged.

**Comment by rgommers** [comment]
> However, @rgommers was suggesting the ufunc implementation in the issue [here](https://github.com/scipy/scipy/issues/17890#issuecomment-1410100391) so let's get his input.

I think I only responded to a question which already presumed needing to write this algorithm in C/Cython. If pure Python works well, that's always much better.

**Pull request closed without merging**
